adapters/sonobi/sonobi.go

```diff
@@ -0,0 +1,296 @@
+			break
+		}
+		sonobiReq.Imp[i].TagID = params.TagID
+	}
+
+	reqJSON, err := json.Marshal(sonobiReq)
+	if err != nil {
+		return nil, err
+	}
+
+	debug := &pbs.BidderDebug{
+		RequestURI: a.URI,
+	}
+	httpReq, _ := http.NewRequest("POST", a.URI, bytes.NewReader(reqJSON))
+	httpReq.Header.Set("Content-Type", "application/json")
+	if sReq.Device != nil {
+		addHeaderIfNonEmpty(httpReq.Header, "User-Agent", sReq.Device.UA)
+		addHeaderIfNonEmpty(httpReq.Header, "X-Forwarded-For", sReq.Device.IP)
+		addHeaderIfNonEmpty(httpReq.Header, "Accept-Language", sReq.Device.Language)
+		addHeaderIfNonEmpty(httpReq.Header, "DNT", strconv.Itoa(int(sReq.Device.DNT)))
+	}
+	sResp, err := ctxhttp.Do(ctx, a.http.Client, httpReq)
+	if err != nil {
+		return nil, err
+	}
+	defer sResp.Body.Close()
+	debug.StatusCode = sResp.StatusCode
+	if sResp.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	body, err := ioutil.ReadAll(sResp.Body)
+	if err != nil {
+		return nil, err
+	}
+	responseBody := string(body)
+	if sResp.StatusCode == http.StatusBadRequest {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("HTTP status %d; body: %s", sResp.StatusCode, responseBody),
+		}
+	}
+
+	if sResp.StatusCode != http.StatusOK {
+		return nil, &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("HTTP status %d; body: %s", sResp.StatusCode, responseBody),
+		}
+	}
+	if req.IsDebug {
+		debug.RequestBody = string(reqJSON)
+		bidder.Debug = append(bidder.Debug, debug)
+		debug.ResponseBody = responseBody
+	}
+
+	var bidResp openrtb.BidResponse
+	err = json.Unmarshal(body, &bidResp)
+	if err != nil {
+		return nil, &errortypes.BadServerResponse{
+			Message: err.Error(),
+		}
+	}
+	bids := make(pbs.PBSBidSlice, 0)
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			bidID := bidder.LookupBidID(bid.ImpID)
+			if bidID == "" {
+				return nil, &errortypes.BadServerResponse{
+					Message: fmt.Sprintf("Unknown ad unit code '%s'", bid.ImpID),
+				}
+			}
+
+			adm, _ := url.QueryUnescape(bid.AdM)
+			pbid := pbs.PBSBid{
+				BidID:       bidID,
+				AdUnitCode:  bid.ImpID,
+				BidderCode:  bidder.BidderCode,
+				Price:       bid.Price,
+				Adm:         adm,
+				Creative_id: bid.CrID,
+				Width:       bid.W,
+				Height:      bid.H,
+				DealId:      bid.DealID,
+				NURL:        bid.NURL,
+			}
+			bids = append(bids, &pbid)
+		}
+	}
+
+	sort.Sort(bids)
+	return bids, nil
+}
+
+// MakeRequests Makes the OpenRTB request payload
+func (a *SonobiAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var sonobiExt openrtb_ext.ExtImpSonobi
+	var err error
+
+	var adapterRequests []*adapters.RequestData
+
+	for _, imp := range request.Imp {
+		// Sonobi doesn't allow multi-type imp. Banner takes priority over video.
+		if imp.Banner != nil {
+		} else if imp.Video != nil {
+		} else {
+			err := fmt.Errorf("Sonobi only supports banner and video imps. Ignoring imp id=%s", imp.ID)
```

Comment:
The base PBS should now take care of this for you. It will not pass any media type objects you don't support, and drop the particular imp if nothing is left that you do.

Reply:
In case it's relevant, note that this instance of "err" (line 185, 186) is not the same as the one declared at the top of the function (line 176), which is the one used at lines 195 and 200, below.  I don't think there's anything technically incorrect here, but it's potentially confusing for future maintenance.  Perhaps renaming this one, or just changing the ":=" to "=" would be less risky!?

Reply:
Must have been left over code from a while back when the first efforts to make this adapter was being made. I removed the check for media types since they will be filtered to the ones that Sonobi supports before hand.